disk-buffering/src/main/java/io/opentelemetry/contrib/disk/buffering/internal/storage/files/ReadableFile.java

```diff
@@ -0,0 +1,153 @@
+import io.opentelemetry.contrib.disk.buffering.internal.storage.files.reader.ReadResult;
+import io.opentelemetry.contrib.disk.buffering.internal.storage.files.reader.StreamReader;
+import io.opentelemetry.contrib.disk.buffering.internal.storage.files.utils.FileTransferUtil;
+import io.opentelemetry.contrib.disk.buffering.internal.storage.responses.ReadableResult;
+import io.opentelemetry.contrib.disk.buffering.internal.storage.utils.StorageClock;
+import java.io.BufferedInputStream;
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.function.Function;
+import javax.annotation.Nullable;
+
+public final class ReadableFile extends StorageFile {
+  private final int originalFileSize;
+  private final StreamReader reader;
+  private final FileTransferUtil fileTransferUtil;
+  private final File temporaryFile;
+  private final StorageClock clock;
+  private final long expireTimeMillis;
+  private final AtomicBoolean isClosed = new AtomicBoolean(false);
+  private int readBytes = 0;
+  @Nullable private ReadResult unconsumedResult;
+
+  public ReadableFile(
+      File file, long createdTimeMillis, StorageClock clock, StorageConfiguration configuration)
+      throws IOException {
+    this(
+        file,
+        createdTimeMillis,
+        clock,
+        configuration,
+        DelimitedProtoStreamReader.Factory.getInstance());
+  }
+
+  public ReadableFile(
+      File file,
+      long createdTimeMillis,
+      StorageClock clock,
+      StorageConfiguration configuration,
+      StreamReader.Factory readerFactory)
+      throws IOException {
+    super(file);
+    this.clock = clock;
+    expireTimeMillis = createdTimeMillis + configuration.getMaxFileAgeForReadMillis();
+    originalFileSize = (int) file.length();
+    temporaryFile = configuration.getTemporaryFileProvider().createTemporaryFile(file.getName());
+    copyFile(file, temporaryFile);
+    FileInputStream tempInputStream = new FileInputStream(temporaryFile);
+    fileTransferUtil = new FileTransferUtil(tempInputStream, file);
+    reader = readerFactory.create(tempInputStream);
+  }
+
+  /**
+   * Reads the next line available in the file and provides it to a {@link Function processing}
+   * which will determine whether to remove the provided line or not.
+   *
+   * @param processing - A function that receives the line that has been read and returns a boolean.
+   *     If the processing function returns TRUE, then the provided line will be deleted from the
+   *     source file. If the function returns FALSE, no changes will be applied to the source file.
+   */
+  public synchronized ReadableResult readAndProcess(Function<byte[], Boolean> processing)
+      throws IOException {
+    if (isClosed.get()) {
+      return ReadableResult.FAILED;
+    }
+    if (hasExpired()) {
+      close();
```

Comment:
something else is supposed to delete the expired file?

Reply:
A clean up is done when creating a new file [here](https://github.com/open-telemetry/opentelemetry-java-contrib/pull/913/files#diff-5132fe2f62753f8cdbc84a13763f484ddbeaea87f8af84fcc8519bdbbe213f5eR84).